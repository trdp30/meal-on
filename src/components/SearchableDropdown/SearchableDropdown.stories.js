/**
 *
 * Stories for SearchableDropdown
 *
 *
 */

import React from 'react';
import SearchableDropdown from './index';
import { defaultProps, search, items } from './tests/mocks';

export default {
  /* ðŸ‘‡ The title prop is optional.
   * See https://storybook.js.org/docs/react/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  title: 'Components/SearchableDropdown',
  component: SearchableDropdown,
  decorators: [Story => <Story />],
};

const Template = args => <SearchableDropdown {...args} />;

export const Basic = Template.bind({});

Basic.args = {
  ...defaultProps,
  search,
  setSelected: () => true,
  selectedItems: [],
  searchFn: (searchText, callback) => callback(items),
};
