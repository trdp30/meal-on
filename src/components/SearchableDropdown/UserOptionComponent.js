import React, { memo } from 'react';
import PropTypes from 'prop-types';
import User from 'components/base/UserDetails/user';

const UserOptionComponent = props => {
  const { option } = props;
  if (option && option.user) {
    return <User data={option.user} size="xs" border="border-none" />;
  }
  return <></>;
};

UserOptionComponent.propTypes = {
  option: PropTypes.shape({
    user: PropTypes.shape({
      id: PropTypes.number.isRequired,
      email: PropTypes.string.isRequired,
    }),
  }).isRequired,
};

export default memo(UserOptionComponent);
