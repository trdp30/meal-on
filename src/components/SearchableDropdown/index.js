/* eslint-disable react/no-unstable-nested-components */
import React, { memo, useMemo, useRef } from 'react';
import PropTypes from 'prop-types';
import AsyncSelect from 'react-select/async';
import { get } from 'lodash';
import fetchData from 'utils/fetchData';
import { components } from 'react-select';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronDown } from '@fortawesome/pro-solid-svg-icons';
import classNames from 'classnames';
import { faSearch } from '@fortawesome/pro-light-svg-icons';
import CustomOption from './Option';

export const reactSelectDefaultStyles = {
  input: props => ({
    ...props,
    input: {
      boxShadow: 'none !important',
    },
  }),
};

// eslint-disable-next-line no-shadow
function DropdownIndicator(props) {
  if (props.selectProps.isSearchIcon) {
    return (
      <components.DropdownIndicator {...props}>
        <FontAwesomeIcon icon={faSearch} className="text-gray-600 w-[18px] h-[18px]" />
      </components.DropdownIndicator>
    );
  }
  return (
    <components.DropdownIndicator {...props}>
      <FontAwesomeIcon icon={faChevronDown} className="text-gray-600" />
    </components.DropdownIndicator>
  );
}

function RemoteDropdown(props) {
  const {
    placeholder,
    labelKey,
    isMulti = false,
    primaryKey,
    selectedItems,
    setSelected,
    emptyLabel,
    OptionComponent,
    value,
    isDisabled,
    query,
    controlShouldRenderValue = true,
    isClearable,
    customFormatOptionLabel,
    customStyles = {},
    isSearchIcon = false,
  } = props;
  const { queryString, queryKey, queryVariables = {} } = query;

  const loadComponent = useMemo(() => queryString && queryKey, [query]);

  const formatOptionLabel = option => get(option, labelKey, option);
  const getOptionValue = option => get(option, primaryKey, option);

  const handleChange = data => {
    setSelected(data);
  };
  const timer = useRef(null);

  const loadOptions = (searchText, callback) => {
    clearTimeout(timer.current);
    timer.current = setTimeout(async () => {
      clearTimeout(timer.current);
      const response = await fetchData({
        queryString,
        queryKey,
        queryVariables: {
          searchString: searchText ? `%${searchText}%` : '%%',
          ...queryVariables,
        },
      });
      callback(response);
    }, 500);
  };

  const noOptionsMessage = () => emptyLabel;

  if (loadComponent) {
    return (
      <AsyncSelect
        onChange={handleChange}
        value={value}
        isSearchIcon={isSearchIcon}
        isClearable={isClearable}
        isDisabled={isDisabled}
        isSearchable
        isMulti={isMulti}
        cacheOptions
        placeholder={placeholder}
        loadOptions={loadOptions}
        defaultOptions
        noOptionsMessage={noOptionsMessage}
        selectedOptions={selectedItems}
        primaryKey={primaryKey}
        labelKey={labelKey}
        OptionComponent={OptionComponent}
        components={{ Option: CustomOption, DropdownIndicator }}
        controlShouldRenderValue={controlShouldRenderValue}
        formatOptionLabel={customFormatOptionLabel || formatOptionLabel}
        styles={{ ...reactSelectDefaultStyles, ...customStyles }}
        getOptionValue={getOptionValue}
        className="chromatic-ignore"
        classNames={{
          control: ({ isFocused }) =>
            classNames(
              'border rounded px-1.5 !shadow-0 !bg-gray-50',
              isFocused ? '!border-blue-300' : '!border-gray-200',
            ),
          placeholder: () => 'text-sm font-normal leading-1 !mx-0 !text-gray-400',
          singleValue: () => 'text-sm font-normal leading-1 !mx-0 !text-gray-900',
          valueContainer: () => '!p-0',
          input: () => '!m-0 !shadow-0',
          indicatorContainer: () => '!p-0',
          indicatorSeparator: () => '!w-0',
          menuList: () => '!p-2 flex flex-col gap-1.5',
          option: ({ isSelected }) =>
            classNames(
              '!px-1.5 !py-2 !text-sm',
              isSelected ? '!bg-blue-50 !text-blue-700' : 'hover:!bg-gray-50 !bg-white',
            ),
        }}
      />
    );
  }
  return <></>;
}

RemoteDropdown.propTypes = {
  OptionComponent: PropTypes.any,
  value: PropTypes.any,
  setSelected: PropTypes.func,
  placeholder: PropTypes.string,
  isMulti: PropTypes.bool,
  selectedItems: PropTypes.any,
  primaryKey: PropTypes.string,
  isDisabled: PropTypes.bool,
  emptyLabel: PropTypes.any,
  query: PropTypes.shape({
    queryString: PropTypes.string.isRequired,
    queryKey: PropTypes.string.isRequired,
    queryVariables: PropTypes.object,
  }).isRequired,
  isClearable: PropTypes.bool,
  customFormatOptionLabel: PropTypes.func,
  controlShouldRenderValue: PropTypes.bool,
};

RemoteDropdown.defaultProps = {
  isClearable: false,
};

export default memo(RemoteDropdown);
