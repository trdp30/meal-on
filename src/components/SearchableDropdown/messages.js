/*
 * SearchableDropdown Messages
 *
 * This contains all the text for the SearchableDropdown component.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.components.SearchableDropdown';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the SearchableDropdown component!',
  },
});
