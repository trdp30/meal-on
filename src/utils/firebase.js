// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";
import { getAnalytics } from "firebase/analytics";
import { getAuth } from "firebase/auth";

const firebaseConfig = {
  apiKey: process.env.REACT_APP_FIREBASE_API_KEY,
  authDomain: process.env.REACT_APP_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.REACT_APP_FIREBASE_PROJECT_ID,
  storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.REACT_APP_FIREBASE_APP_ID,
  measurementId: process.env.REACT_APP_FIREBASE_MEASUREMENT_ID,
};

let app, analytics, auth;

// Initialize Firebase
// export default async function initializeFirebase() {
//   app = await initializeApp(firebaseConfig, "menuOnRoad");
//   analytics = await getAnalytics(app);

//   if (process.env.NODE_ENV !== "production") {
//     // analytics.app.disable();
//   }
//   auth = await getAuth(app);
//   return {
//     app,
//     analytics,
//     auth,
//   };
// }

app = initializeApp(firebaseConfig, "menuOnRoad");
analytics = getAnalytics(app);

if (process.env.NODE_ENV !== "production") {
  // analytics.app.disable();
}
auth = getAuth(app);

console.log("auth", auth);

export { app, analytics, auth };

export default app;
