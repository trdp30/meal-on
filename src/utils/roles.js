const roles = { admin: "ADMIN", owner: "OWNER", manager: "MANAGER" };

export default roles;
