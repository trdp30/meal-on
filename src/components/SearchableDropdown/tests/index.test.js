/**
 *
 * Tests for SearchableDropdown
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import fetchData from 'utils/fetchData';
import userEvent from '@testing-library/user-event';
import { render, waitFor } from 'utils/testHelpers';
import '@testing-library/jest-dom';
import Auth from 'utils/auth';
import { defaultProps, items } from './mocks';
import SearchableDropdown from '../index';

jest.mock('utils/auth');
// beforeEach(() => {
//   jest.useFakeTimers();
// });

// beforeEach(() => {
//   jest.runAllTimers();
// });

Auth.client.mockImplementation(() => ({ query: jest.fn().mockResolvedValue({ data: items }) }));

global.fetch = jest.fn(() => Promise.resolve());

jest.mock('lodash', () => ({
  ...jest.requireActual('lodash'),
  debounce: fn => {
    // eslint-disable-next-line no-param-reassign
    fn.cancel = jest.fn();
    return fn;
  },
}));

jest.mock('utils/fetchData');
fetchData.mockImplementation(() => [
  { id: 1, name: 'Option 1' },
  { id: 2, name: 'Option 2' },
]);

describe('<SearchableDropdown />', () => {
  it('Expect to not log errors in console', async () => {
    const spy = jest.spyOn(global.console, 'error');
    const onChange = jest.fn();
    const searchFn = jest.fn().mockReturnValue([]);
    const selectedItems = [];
    const { getByRole } = render(
      <SearchableDropdown {...defaultProps} setSelected={onChange} selectedItems={selectedItems} search={searchFn} />,
    );
    await waitFor(() => getByRole('combobox'));
    expect(spy).not.toHaveBeenCalled();
    spy.mockReset();
    spy.mockRestore();
  });

  it('Should render and match the snapshot', async () => {
    const onChange = jest.fn();
    const searchFn = jest.fn().mockReturnValue([]);
    const selectedItems = [];
    const { container, getByRole } = render(
      <SearchableDropdown {...defaultProps} setSelected={onChange} selectedItems={selectedItems} search={searchFn} />,
    );
    await waitFor(() => getByRole('combobox'));
    expect(container).toMatchSnapshot();
  });
  const props = {
    placeholder: 'Select an option',
    labelKey: 'name',
    primaryKey: 'id',
    selectedItems: [],
    setSelected: jest.fn(),
    emptyLabel: 'No options found',
    query: {
      queryString: '/api/options',
      queryKey: 'search',
      queryVariables: {},
    },
  };

  it('renders correctly', () => {
    const { getByText } = render(<SearchableDropdown {...props} />);
    const dropdown = getByText('Select an option');
    expect(dropdown).toBeInTheDocument();
  });

  it('fetch Data is called upon loading the component', async () => {
    render(<SearchableDropdown {...props} />);
    await waitFor(() => expect(fetchData).toBeCalled());
  });

  it('options loaded on typing ', async () => {
    const { container, getByText } = render(<SearchableDropdown {...props} />);
    await waitFor(() => expect(fetchData).toBeCalled());
    const user = userEvent.setup();
    const dropdown = getByText('Select an option');
    await user.type(dropdown, 'Option 1');
    expect(container).toMatchSnapshot();
  });

  // it('options loaded on typing ', async () => {
  //   jest.useFakeTimers();
  //   const { getByText, getByRole } = render(<SearchableDropdown {...props} />);
  //   jest.runAllTimers();
  //   const user = userEvent.setup();
  //   const dropdown = getByText('Select an option');
  //   await user.click(dropdown);
  //   jest.runAllTimers();
  //   const dropdownInput = getByRole('textbox');
  //   console.log('dropdownInput', dropdownInput);
  //   await user.type(dropdownInput, '1');
  //   jest.runAllTimers();

  //   screen.debug();
  //   await user.click(getByText('Option 1'));
  // });

  // it('should render the options on the on clicking the combobox button', async () => {
  //   const user = userEvent.setup();
  //   const onChange = jest.fn();
  //   const searchFn = jest.fn().mockReturnValue(items);

  //   const selectedItems = [];
  //   const { container, getByRole, getAllByRole } = render(
  //     <SearchableDropdown {...defaultProps} onChange={onChange} selectedItems={selectedItems} search={searchFn} />,
  //   );
  //   await waitFor(() => getByRole('button'));
  //   await user.click(getByRole('button'));
  //   expect(searchFn).toHaveBeenCalled();

  //   await waitFor(() => getAllByRole('option'));
  //   expect(getAllByRole('option')).toHaveLength(items.length);
  //   expect(container).toMatchSnapshot();
  // });

  // it('should have empty options while search function is not provided as prop', async () => {
  //   const user = userEvent.setup();
  //   const onChange = jest.fn();
  //   const spy = jest.spyOn(global.console, 'error');

  //   const selectedItems = [];
  //   const { container, getByRole, getByText } = render(
  //     <SearchableDropdown {...defaultProps} onChange={onChange} selectedItems={selectedItems} />,
  //   );
  //   await user.click(getByRole('button'));

  //   await waitFor(() => getByRole('option'));
  //   expect(getByText('No results')).toBeTruthy();
  //   expect(container).toMatchSnapshot();
  //   expect(spy).toHaveBeenCalled();
  //   spy.mockReset();
  //   spy.mockRestore();
  // });

  // it('should have empty options while queryString is not provided as prop', async () => {
  //   const user = userEvent.setup();
  //   const onChange = jest.fn();
  //   const searchFn = jest.fn().mockReturnValue(items);

  //   const spy = jest.spyOn(global.console, 'error');

  //   const selectedItems = [];
  //   const { ...props } = defaultProps;
  //   const { getByRole, getByText } = render(
  //     <SearchableDropdown {...props} onChange={onChange} selectedItems={selectedItems} search={searchFn} />,
  //   );
  //   await user.click(getByRole('button'));

  //   await waitFor(() => getByRole('option'));
  //   expect(getByText('No results')).toBeTruthy();
  //   expect(spy).toHaveBeenCalled();
  //   spy.mockReset();
  //   spy.mockRestore();
  // });

  // it('should have empty options while queryKey is not provided as prop', async () => {
  //   const user = userEvent.setup();
  //   const onChange = jest.fn();
  //   const searchFn = jest.fn().mockReturnValue(items);
  //   const spy = jest.spyOn(global.console, 'error');

  //   const selectedItems = [];
  //   const { queryKey, ...props } = defaultProps;
  //   const { getByRole, getByText } = render(
  //     <SearchableDropdown {...props} onChange={onChange} selectedItems={selectedItems} search={searchFn} />,
  //   );
  //   await user.click(getByRole('button'));

  //   await waitFor(() => getByRole('option'));
  //   expect(getByText('No results')).toBeTruthy();
  //   expect(spy).toHaveBeenCalled();
  //   spy.mockReset();
  //   spy.mockRestore();
  // });

  // it('should call the search function when data entered in the search box', async () => {
  //   const user = userEvent.setup();
  //   const onChange = jest.fn();
  //   const searchFn = jest.fn().mockReturnValue(items);
  //   const spy = jest.spyOn(global.console, 'error');

  //   const selectedItems = [];
  //   const { getByRole } = render(
  //     <SearchableDropdown {...defaultProps} onChange={onChange} selectedItems={selectedItems} search={searchFn} />,
  //   );
  //   await user.click(getByRole('button'));
  //   await user.type(getByRole('combobox'), 'a');
  //   expect(searchFn).toBeCalledTimes(2);
  //   expect(spy).not.toHaveBeenCalled();
  //   spy.mockReset();
  //   spy.mockRestore();
  // });

  // it('should have empty while providing invalid queryString', async () => {
  //   const user = userEvent.setup();
  //   const onChange = jest.fn();
  //   const searchFn = jest.fn().mockReturnValue(items);

  //   const selectedItems = [];
  //   const { getByText, getByRole } = render(
  //     <SearchableDropdown
  //       {...defaultProps}
  //       queryString="hello world"
  //       onChange={onChange}
  //       selectedItems={selectedItems}
  //       search={searchFn}
  //     />,
  //   );
  //   await user.click(getByRole('button'));
  //   await waitFor(() => getByRole('option'));
  //   expect(getByText('No results')).toBeTruthy();
  // });
});
