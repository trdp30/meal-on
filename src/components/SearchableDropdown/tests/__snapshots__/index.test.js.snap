// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchableDropdown /> Should render and match the snapshot 1`] = `
<div>
  <div
    class="chromatic-ignore css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-3-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class="border rounded px-1.5 !shadow-0 !bg-gray-50 !border-gray-200 css-13cymwt-control"
    >
      <div
        class="!p-0 css-1fdsijx-ValueContainer"
      >
        <div
          class="text-sm font-normal leading-1 !mx-0 !text-gray-400 css-1jqq78o-placeholder"
          id="react-select-3-placeholder"
        >
          Search User
        </div>
        <div
          class="!m-0 !shadow-0 css-ieh8vb-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-describedby="react-select-3-placeholder"
            aria-expanded="false"
            aria-haspopup="true"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class=""
            id="react-select-3-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class=" css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class=" css-pogzpp-loadingIndicator"
        >
          <span
            class="css-1xtdfmb-LoadingDot"
          />
          <span
            class="css-zoievk-LoadingDot"
          />
          <span
            class="css-x748d8-LoadingDot"
          />
        </div>
        <span
          class="!w-0 css-1u9des2-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class=" css-1xc3v61-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down text-gray-600"
            data-icon="chevron-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SearchableDropdown /> options loaded on typing  1`] = `
<div>
  <div
    class="chromatic-ignore css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-6-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    >
      <span
        id="aria-selection"
      />
      <span
        id="aria-context"
      >
          Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
      </span>
    </span>
    <div
      class="border rounded px-1.5 !shadow-0 !bg-gray-50 !border-blue-300 css-t3ipsp-control"
    >
      <div
        class="!p-0 css-1fdsijx-ValueContainer"
      >
        <div
          class="!m-0 !shadow-0 css-zoi0k-Input"
          data-value="Option 1"
        >
          <input
            aria-autocomplete="list"
            aria-controls="react-select-6-listbox"
            aria-describedby="react-select-6-placeholder"
            aria-expanded="true"
            aria-haspopup="true"
            aria-owns="react-select-6-listbox"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class=""
            id="react-select-6-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value="Option 1"
          />
        </div>
      </div>
      <div
        class=" css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class=" css-o9ehiq-loadingIndicator"
        >
          <span
            class="css-1xtdfmb-LoadingDot"
          />
          <span
            class="css-zoievk-LoadingDot"
          />
          <span
            class="css-x748d8-LoadingDot"
          />
        </div>
        <span
          class="!w-0 css-1u9des2-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class=" css-15lsz6c-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down text-gray-600"
            data-icon="chevron-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class=" css-1nmdiq5-menu"
      id="react-select-6-listbox"
    >
      <div
        class="!p-2 flex flex-col gap-1.5 css-1n6sfyn-MenuList"
      >
        <div
          class=" css-zu7x0v-LoadingMessage"
        >
          Loading...
        </div>
      </div>
    </div>
  </div>
</div>
`;
